src/field255/integer.rs

```diff
@@ -0,0 +1,839 @@
+use std::{
```

Comment:
Before reviewing this code too closely, I think we should first try to answer a basic question: How much of this functionality are we actually going to need?

To answer this question, I played around with the IDPF code (#381) and found that we only use the arithmetic in tests. (I replaced the arithmetic functions with `todo!()` to see what functions panicked; none of the IDPF tests panicked.)

Looking ahead, I don't think we're actually going to need 256-bit arithmetic. Certainly not for Poplar.

In fact, I think the only thing we need this code for is to satisfy the (rather stringent) requirements of the `FieldElement` trait, which demands that each field element has a corresponding integer representation, and that type has to implement a number of arithmetic operations. (We get these for free for rust's primitive types.)

This being the case, I wonder if it would be worth trying to relax the API requirements, rather than provide an implementation of an API we're not likely to use. Specifically:

(1) Remove the `Integer` associated type from `FieldElement` 

(2) Define a new trait for `FieldElement`s that have an integer representation. Something like:

```rust
trait IntoInteger: FieldElement {
    type Integer: Add + Sub + Mul + ...;

    // we would need to move this function here, as well any other trait method that operates on the integer representation
    fn modulus() -> Self::Integer;
}
```

Let me know what you think  ... I might be missing something important here.

Reply:
I implemented the change to add FromInteger, but as I was reaching for num-bigint to reimplement `Display for Field255`, I realized we could just use `num_bigint::BigInt` as `Field255::Integer`. Thoughts? I think it should be sufficient for the ancillary conversion uses, plus it would simplify the field element tests if we could keep the `Field::Integer` associated type.

Reply:
Looking ahead to the IDPF, we will need some way to convert a control bit to a `Field64`/`Field255` element, so that we can use multiplication instead of flow control to apply field element corrections. From this decision point, that could either mean we add such a conversion method to the `FieldElement` trait (or another new trait) or bouncing through two `Into` conversions (u8 -> Field64::Integer=u64 -> Field64, or alternately u8 -> Field255::Integer=BigInt -> Field255).

Reply:
Which tests do we need `Field::Integer` for?

Reply:
The main things are test cases that use the field's modulus to construct interesting elements, (like p-1, or checking what happens when you pass the saturated modulus) and the exponent argument to `pow()`. I had to move some checks from `field_element_test_common()` to `field_element_test()`. The change to `pow()` keeps me from using the field modulus (or p - 1) in a field element exponentiation.

Reply:
> The main things are test cases that use the field's modulus to construct interesting elements, (like p-1, or checking what happens when you pass the saturated modulus) and the exponent argument to pow().

Ah yes, I think that's still good to test. I think going with big_int::BigInt for Field255::Integer is the right way to go, but I would only compile this code in tests (i.e., use tag `#[cfg(tests)]`). The goal being to fence code we don't intend to be used by other applications.

> Looking ahead to the IDPF, we will need some way to convert a control bit to a Field64/Field255 element, ...

Can you remind where this happens? I can't seem to find the relevant bit in the spec ...

Reply:
In `gen()`, there's the following snippet:

```
        if ctrl[1] == Field2(1):
            w_cw = vec_neg(w_cw)
```

And in `eval()`:

```
    if next_ctrl == Field2(1):
        y = vec_add(y, w_cw)
```

These branches could be eliminated with `w_cw ← w_cw * (1 - 2 * Field(ctrl[1]))` and `y ← y + w_cw * Field(next_ctrl)`.

Reply:
Nice. Currently the conversion path is `usize -(TryInto)-> FieldElement::Integer -(Into)-> FieldElement`. WDYTA this simplification:

1. Require `TryFrom<usize>` on `FieldElement`
2. Remove `TryFrom<usize>` from `IntoInteger`.

I don't think (2) is not strictly necessary, but if we're making breaking changes, and this is better, than it's worth considering. I'm not sure how much effort it would be to refactor.